src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileMetadataService.scala

```diff
@@ -4,33 +4,42 @@ import java.sql.{SQLException, Timestamp}
 import java.util.UUID
 
 import uk.gov.nationalarchives.Tables.{FilemetadataRow, FilepropertyRow}
-import uk.gov.nationalarchives.tdr.api.db.repository.{FileMetadataRepository, FilePropertyRepository}
+import uk.gov.nationalarchives.tdr.api.db.repository.{ClientFileMetadataRepository, FileMetadataRepository, FilePropertyRepository, FileRepository}
 import uk.gov.nationalarchives.tdr.api.graphql.DataExceptions.InputDataException
 import uk.gov.nationalarchives.tdr.api.graphql.fields.FileMetadataFields.{AddFileMetadataInput, FileMetadata}
+import uk.gov.nationalarchives.tdr.api.graphql.fields.FileMetadataFields.SHA256ServerSideChecksum
 
 import scala.concurrent.{ExecutionContext, Future}
 
 class FileMetadataService(fileMetadataRepository: FileMetadataRepository, filePropertyRepository: FilePropertyRepository,
+                          clientFileMetadataService: ClientFileMetadataService,
                           timeSource: TimeSource, uuidSource: UUIDSource)(implicit val ec: ExecutionContext) {
 
   def addFileMetadata(addFileMetadataInput: AddFileMetadataInput, userId: Option[UUID]): Future[FileMetadata] = {
-    getFileProperty(addFileMetadataInput.filePropertyName).flatMap {
-      case Some(property) =>
-        val row = FilemetadataRow(uuidSource.uuid,
-          addFileMetadataInput.fileId,
-          property.propertyid,
-          addFileMetadataInput.value,
-          Timestamp.from(timeSource.now),
-          userId.get)
-
-        fileMetadataRepository.addFileMetadata(row).map(r => FileMetadata(property.name.get, r.fileid, r.value)).recover {
-          case e: SQLException => throw InputDataException(e.getLocalizedMessage, Some(e))
+    def row(property: FilepropertyRow) = FilemetadataRow(uuidSource.uuid,
+      addFileMetadataInput.fileId,
+      property.propertyid,
+      addFileMetadataInput.value,
+      Timestamp.from(timeSource.now),
+      userId.get)
+    val filePropertyName = addFileMetadataInput.filePropertyName
+
+    filePropertyName match {
+      case SHA256ServerSideChecksum =>
+        (for {
+          fileProperty <- getFileProperty(addFileMetadataInput.filePropertyName)
+          cfm <- clientFileMetadataService.getClientFileMetadata(addFileMetadataInput.fileId) if fileProperty.isDefined
+          row <- fileMetadataRepository.addChecksumMetadata(row(fileProperty.get), cfm.checksum.map(_ == addFileMetadataInput.value))
+        } yield FileMetadata(fileProperty.flatMap(_.name).get, row.fileid, row.value)) recover {
+          case e: Throwable => throw InputDataException(e.getLocalizedMessage, Some(e))
         }
-      case None => throw InputDataException(s"The property does not exist", Option.empty)
+      case _ => Future.failed(InputDataException(s"$filePropertyName found. We are only expecting checksum updates for now"))
```

Comment:
Not worth it now (given that we'll probably have to reshuffle the metadata code either for the auditing or the custom metadata anyway), but it might be good to see whether Sangria plays nicely with [GraphQL enums](https://graphql.org/learn/schema/#enumeration-types). Maybe we won't need this error handling at all if we can reject invalid property names at the schema level?

Reply:
That would be good. Worth looking at after we've finished with the metadata code then.